File: app/src/main/java/com/scatl/uestcbbs/module/board/view/BoardActivity.java
package com.scatl.uestcbbs.module.board.view;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.viewpager2.widget.ViewPager2;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;
import com.jaeger.library.StatusBarUtil;
import com.luck.picture.lib.PictureSelector;
import com.luck.picture.lib.config.PictureConfig;
import com.luck.picture.lib.config.PictureMimeType;
import com.luck.picture.lib.entity.LocalMedia;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.annotation.ToastType;
import com.scatl.uestcbbs.base.BaseActivity;
import com.scatl.uestcbbs.base.BasePresenter;
import com.scatl.uestcbbs.entity.ForumDetailBean;
import com.scatl.uestcbbs.entity.SingleBoardBean;
import com.scatl.uestcbbs.entity.SubForumListBean;
import com.scatl.uestcbbs.helper.glidehelper.GlideEngineForPictureSelector;
import com.scatl.uestcbbs.module.board.adapter.BoardPostViewPagerAdapter;
import com.scatl.uestcbbs.module.board.presenter.BoardPresenter;
import com.scatl.uestcbbs.util.ExtensionKt;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.FileUtil;
import com.scatl.uestcbbs.util.ImageUtil;
import com.scatl.uestcbbs.util.SharePrefUtil;

import java.util.ArrayList;
import java.util.List;

public class BoardActivity extends BaseActivity implements BoardView, AppBarLayout.OnOffsetChangedListener {

    private AppBarLayout appBarLayout;
    private CollapsingToolbarLayout collapsingToolbarLayout;
    private Toolbar toolbar;
    private CoordinatorLayout coordinatorLayout;
    private View boardInfoLayout;
    private ProgressBar progressBar;
    private TextView hint, postsNumTv;
    private ImageView boardIcon;
    private ImageView boardBackground;
    private TabLayout mTabLayout;
    private ViewPager2 viewPager;
    private TextView boardNameTv;

    private BoardPresenter boardPresenter;
    private int boardId;
    private String boardName;

    private SubForumListBean subForumListBean;
    private SingleBoardBean singleBoardBean;

    private static final int ACTION_SELECT_PHOTO = 99;
    private static final int ACTION_MODIFY_PHOTO = 100;

    @Override
    protected void getIntent(Intent intent) {
        super.getIntent(intent);
        boardId = intent.getIntExtra(Constant.IntentKey.BOARD_ID, Integer.MAX_VALUE);
        boardName = intent.getStringExtra(Constant.IntentKey.BOARD_NAME);
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_board;
    }

    @Override
    protected void findView() {
        collapsingToolbarLayout = findViewById(R.id.board_toolbar_layout);
        progressBar = findViewById(R.id.board_progressbar);
        hint = findViewById(R.id.board_hint);
        appBarLayout = findViewById(R.id.board_app_bar);
        coordinatorLayout = findViewById(R.id.board_coor_layout);
        toolbar = findViewById(R.id.toolbar);
        boardInfoLayout = findViewById(R.id.board_info_layout);
        boardIcon = findViewById(R.id.board_icon);
        boardBackground = findViewById(R.id.board_background);
        mTabLayout = findViewById(R.id.board_indicator);
        viewPager = findViewById(R.id.board_viewpager);
        boardNameTv = findViewById(R.id.board_name);
        postsNumTv = findViewById(R.id.board_posts_num);
    }

    @Override
    protected void initView() {
        super.initView();
        boardPresenter = (BoardPresenter) presenter;

        ExtensionKt.desensitize(viewPager);
        boardNameTv.setText(boardName);

        appBarLayout.addOnOffsetChangedListener(this);
        boardBackground.setOnClickListener(this::onClickListener);

        setSupportActionBar(toolbar);

        loadBoardImg();
        boardPresenter.getForumDetail(this, boardId);
        if(boardId == Constant.DEPARTMENT_BOARD_ID) {
            String data = FileUtil.readAssetFile(this, "department.json");
            if (JSON.isValidObject(data)) {
                try {
                    JSONObject jsonObject = JSONObject.parseObject(data);
                    SubForumListBean subForumListBean = JSON.toJavaObject(jsonObject, SubForumListBean.class);
                    onGetSubBoardListSuccess(subForumListBean);
                } catch (Exception e) {
                    showToast("出错了，请联系开发者", ToastType.TYPE_ERROR);
                }
            }
        } else {
            boardPresenter.getSubBoardList(boardId, this);
        }
    }

    @Override
    protected BasePresenter initPresenter() {
        return new BoardPresenter();
    }

    @Override
    protected void onClickListener(View view) {
        if (view.getId() == R.id.board_background) {
            PictureSelector.create(this)
                    .openGallery(PictureMimeType.ofImage())
                    .isCamera(false)
                    .isGif(false)
                    .showCropFrame(true)
                    .hideBottomControls(false)
                    .theme(R.style.picture_WeChat_style)
                    .maxSelectNum(1)
                    .isEnableCrop(true)
                    .withAspectRatio(3, 2)
                    .imageEngine(GlideEngineForPictureSelector.createGlideEngine())
                    .forResult(PictureConfig.CHOOSE_REQUEST);
        }
    }

    private void loadBoardImg() {
        Glide
                .with(this)
                .load(SharePrefUtil.getBoardImg(this, boardId))
                .into(boardIcon);

        try {
            Glide
                    .with(this)
                    .load(SharePrefUtil.getBoardImg(this, boardId))
                    .into(new SimpleTarget<Drawable>() {
                        @Override
                        public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
                            Bitmap bitmap = resource instanceof GifDrawable ? ((GifDrawable) resource).getFirstFrame() : ImageUtil.drawable2Bitmap(resource);
                            boardBackground.setImageBitmap(ImageUtil.blurPhoto(BoardActivity.this, bitmap, 25));
                        }
                    });
        } catch (Exception e) {
//            Glide.with(this)
//                    .load(SharePrefUtil.getBoardImg(this, boardId))
//                    .apply(new RequestOptions().transform(new BlurTransformation()))
//                    .into(boardBackground);
        }
    }

    @Override
    public void onGetSubBoardListSuccess(SubForumListBean subForumListBean) {
        this.subForumListBean = subForumListBean;
        AlphaAnimation alphaAnimation = new AlphaAnimation(0, 1);
        alphaAnimation.setDuration(600);
        coordinatorLayout.startAnimation(alphaAnimation);
        coordinatorLayout.setVisibility(View.VISIBLE);
        progressBar.setVisibility(View.GONE);
        hint.setVisibility(View.GONE);

        List<Integer> ids = new ArrayList<>();
        String[] titles;
        ids.add(boardId);

        if (subForumListBean.list != null && subForumListBean.list.size() != 0) {

            postsNumTv.setText("今日：" + subForumListBean.list.get(0).td_posts_num + " | " +
                    "主题：" + subForumListBean.list.get(0).topic_total_num + " | " +
                    "帖子：" + subForumListBean.list.get(0).posts_total_num);
            boardNameTv.setText(subForumListBean.list.get(0).board_category_name);

            titles = new String[subForumListBean.list.get(0).board_list.size() + 1];
            titles[0] = boardName;

            for (int i = 0; i < subForumListBean.list.get(0).board_list.size(); i ++) {
                titles[i + 1] = subForumListBean.list.get(0).board_list.get(i).board_name;
                ids.add(subForumListBean.list.get(0).board_list.get(i).board_id);
            }
        } else {
            titles = new String[1];
            titles[0] = boardName;
            mTabLayout.setVisibility(View.GONE);
        }

        viewPager.setOffscreenPageLimit(titles.length);
        viewPager.setAdapter(new BoardPostViewPagerAdapter(this, ids));
        viewPager.setCurrentItem(0);
        viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                BoardActivity.this.onPageSelected(position);
            }
        });

        new TabLayoutMediator(mTabLayout, viewPager, new TabLayoutMediator.TabConfigurationStrategy() {
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                tab.setText(titles[position]);
            }
        }).attach();

    }

    @Override
    public void onGetSubBoardListError(String msg) {
        progressBar.setVisibility(View.GONE);
        hint.setText(msg);
    }


    @Override
    public void onPermissionRefusedWithNoMoreRequest() {
        showToast(getString(R.string.permission_refuse), ToastType.TYPE_ERROR);
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        int scrollRange = appBarLayout.getTotalScrollRange();
        float alpha = 1 - (1.0f * (- verticalOffset)) / scrollRange;
        boardInfoLayout.setAlpha(alpha);
        toolbar.setTitle(boardName);
        toolbar.setAlpha(1-alpha);
    }


    @Override
    protected void setStatusBar() {
        super.setStatusBar();
        StatusBarUtil.setTransparent(this);
    }

    @Override
    protected void onOptionsSelected(MenuItem item) {
        super.onOptionsSelected(item);
        if (item.getItemId() == R.id.menu_board_set_background_default) {
            SharePrefUtil.setBoardImg(this, boardId, "file:///android_asset/board_img/" + boardId + ".jpg");
            loadBoardImg();
        }
    }

    @Override
    public void onGetForumDetailSuccess(ForumDetailBean forumDetailBean) {

    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == PictureConfig.CHOOSE_REQUEST) {
            List<LocalMedia> selectList = PictureSelector.obtainMultipleResult(data);
            if (selectList.size() != 0 && selectList.get(0).isCut()) {
                SharePrefUtil.setBoardImg(this, boardId, selectList.get(0).getCutPath());
                loadBoardImg();
            }
        }
    }

    private void onPageSelected(int position) {
        if (subForumListBean != null && ExtensionKt.isNotNullAndEmpty(subForumListBean.list)) {
            SubForumListBean.ListBean listBean = subForumListBean.list.get(0);
            if (ExtensionKt.isNotNullAndEmpty(listBean.board_list)) {
                if (position != 0) {
                    postsNumTv.setText("今日：" + listBean.board_list.get(position - 1).td_posts_num + " | " +
                            "主题：" + listBean.board_list.get(position - 1).topic_total_num + " | " +
                            "帖子：" + listBean.board_list.get(position - 1).posts_total_num);
                    boardNameTv.setText(listBean.board_list.get(position - 1).board_name);
                } else {
                    postsNumTv.setText("今日：" + listBean.td_posts_num + " | " +
                            "主题：" + listBean.topic_total_num + " | " +
                            "帖子：" + listBean.posts_total_num);
                    boardNameTv.setText(listBean.board_category_name);
                }
            }
        }
    }

}

File: app/src/main/java/com/scatl/uestcbbs/module/board/view/BoardActivity.kt
package com.scatl.uestcbbs.module.board.view

import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.view.MenuItem
import android.view.View
import androidx.palette.graphics.Palette
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import com.google.android.material.tabs.TabLayoutMediator
import com.jaeger.library.StatusBarUtil
import com.luck.picture.lib.PictureSelector
import com.luck.picture.lib.config.PictureConfig
import com.luck.picture.lib.config.PictureMimeType
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.annotation.ToastType
import com.scatl.uestcbbs.base.BaseVBActivity
import com.scatl.uestcbbs.databinding.ActivityNewBoardBinding
import com.scatl.uestcbbs.entity.ForumDetailBean
import com.scatl.uestcbbs.entity.SubForumListBean
import com.scatl.uestcbbs.manager.ForumListManager
import com.scatl.uestcbbs.helper.glidehelper.GlideEngineForPictureSelector
import com.scatl.uestcbbs.module.board.adapter.BoardPostViewPagerAdapter
import com.scatl.uestcbbs.module.board.presenter.BoardPresenter
import com.scatl.uestcbbs.module.board.view.behavior.CoverBehavior
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.SharePrefUtil
import com.scatl.uestcbbs.util.isNullOrEmpty
import com.scatl.uestcbbs.util.load
import com.scatl.uestcbbs.util.showToast
import com.scatl.util.ColorUtil
import com.scatl.util.ImageUtil
import com.scatl.util.desensitize
import java.io.File
import java.io.IOException

/**
 * Created by sca_tl at 2023/4/27 10:35
 */
class BoardActivity: BaseVBActivity<BoardPresenter, BoardView, ActivityNewBoardBinding>(),
    BoardView, CoverBehavior.OnCoverViewChanged {

    private var boardId = 0
    private var locateBoardId = 0
    private lateinit var boardName: String
    private var deprecatedBoard = false

    override fun getViewBinding() = ActivityNewBoardBinding.inflate(layoutInflater)

    override fun initPresenter() = BoardPresenter()

    override fun getIntent(intent: Intent?) {
        boardId = intent?.getIntExtra(Constant.IntentKey.BOARD_ID, Int.MAX_VALUE)?:Int.MAX_VALUE
        boardName = intent?.getStringExtra(Constant.IntentKey.BOARD_NAME) ?: ""
        locateBoardId = intent?.getIntExtra(Constant.IntentKey.LOCATE_BOARD_ID, Int.MAX_VALUE)?:Int.MAX_VALUE

        val tmpName = ForumListManager.INSTANCE.getForumInfo(boardId).name ?: ""
        if (tmpName.isNotEmpty()) {
            boardName = tmpName
        }
        if (boardId == 0) {
            deprecatedBoard = true
            boardId = locateBoardId
        }
    }

    override fun initView(theftProof: Boolean) {
        super.initView(true)

        if (boardId == 0 || boardId == Int.MAX_VALUE) {
            showToast("板块不存在！", ToastType.TYPE_ERROR)
            finish()
        }

        mBinding.toolbar.title = boardName
        mBinding.boardName.text = boardName
        mBinding.viewpager.desensitize()

        bindClickEvent(mBinding.coverImg)

        loadBoardImg()
        mPresenter?.getForumDetail(boardId)
        mPresenter?.getSubBoardList(boardId)

        mBinding.statusView.loading(mBinding.coverLayout, mBinding.appBar, mBinding.contentLayout)
    }

    override fun onClick(v: View) {
        if (v == mBinding.coverImg) {
            PictureSelector
                .create(this)
                .openGallery(PictureMimeType.ofImage())
                .isCamera(false)
                .isGif(false)
                .showCropFrame(true)
                .hideBottomControls(false)
                .theme(R.style.picture_WeChat_style)
                .maxSelectNum(1)
                .isEnableCrop(true)
                .withAspectRatio(3, 2)
                .imageEngine(GlideEngineForPictureSelector.createGlideEngine())
                .forResult(PictureConfig.CHOOSE_REQUEST)
        }
    }

    override fun onOptionsSelected(item: MenuItem?) {
        super.onOptionsSelected(item)
        if (item?.itemId == R.id.menu_board_set_background_default) {
            SharePrefUtil.setBoardImg(this, boardId, "file:///android_asset/board_img/$boardId.jpg")
            loadBoardImg()
        }
    }

    override fun onGetSubBoardListSuccess(subForumListBean: SubForumListBean) {
        val ids = mutableListOf(boardId)
        val titles = mutableListOf(boardName)

        if (!subForumListBean.list.isNullOrEmpty()) {
            for (item in subForumListBean.list[0].board_list) {
                titles.add(item.board_name)
                ids.add(item.board_id)
            }
        }

        mBinding.viewpager.apply {
            offscreenPageLimit = titles.size
            adapter = BoardPostViewPagerAdapter(this@BoardActivity, ids)
            currentItem = 0
        }

        TabLayoutMediator(mBinding.tabLayout, mBinding.viewpager) { tab, position ->
            tab.text = titles[position]
        }.attach()

        if (locateBoardId != Int.MAX_VALUE) {
            val locateIndex = ids.indexOf(locateBoardId)
            if (locateIndex in 0 until ids.size) {
                mBinding.viewpager.setCurrentItem(locateIndex, false)
            }
            locateBoardId = Int.MAX_VALUE
        }

        mBinding.statusView.success()
    }

    override fun onGetSubBoardListError(msg: String?) {
        mBinding.statusView.error(msg)
    }

    override fun onGetForumDetailSuccess(forumDetailBean: ForumDetailBean?) {
    }

    private fun getBoardImgPath() = SharePrefUtil.getBoardImg(this, if (deprecatedBoard) 0 else boardId)

    private fun loadBoardImg() {
        mBinding.boardIcon.load(getBoardImgPath())
        setBlurBg()
        try {
            Glide
                .with(this)
                .load(getBoardImgPath())
                .into(object : SimpleTarget<Drawable?>() {
                    override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable?>?) {
                        val bitmap = if (resource is GifDrawable) resource.firstFrame else ImageUtil.drawable2Bitmap(resource)
                        bitmap?.let {
                            mBinding.coverImg.setImageBitmap(ImageUtil.blur(this@BoardActivity, it, 15f))
                        }
                    }
                })
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun setBlurBg() {
        val colors = intArrayOf(
            Color.parseColor("#4D000000"),
            ColorUtil.getAlphaColor(0.6f, Color.parseColor("#4D000000"))
        )

        try {
            if (SharePrefUtil.getBoardImg(getContext(), boardId).contains("android_asset")) {
                getContext()
                    .resources
                    .assets
                    .open(SharePrefUtil.getBoardImg(getContext(), boardId).replace("file:///android_asset/", ""))
            } else {
                File(SharePrefUtil.getBoardImg(getContext(), boardId)).inputStream()
            }.use {
                val bitmap = BitmapFactory.decodeStream(it)
                val palette = Palette.from(bitmap).generate()
                val vibrantSwatch = palette.vibrantSwatch
                val mutedSwatch = palette.mutedSwatch

                if (mutedSwatch != null) {
                    colors[0] = mutedSwatch.rgb
                    colors[1] = ColorUtil.getAlphaColor(0.6f, mutedSwatch.rgb)
                } else if (vibrantSwatch != null) {
                    colors[0] = vibrantSwatch.rgb
                    colors[1] = ColorUtil.getAlphaColor(0.6f, vibrantSwatch.rgb)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

        mBinding.blurImg.background = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, colors)
    }

    override fun getContext() = this

    override fun setStatusBar() {
        StatusBarUtil.setTranslucentForImageView(this, 0, null)
    }

    override fun onCoverChanged(upPercent: Float, downPercent: Float) {
        mBinding.boardIcon.alpha = 1 - upPercent
        mBinding.boardName.alpha = 1 - upPercent
        mBinding.coverImg.alpha = 1 - upPercent
        mBinding.blurImg.alpha = upPercent
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == PictureConfig.CHOOSE_REQUEST) {
            val selectList = PictureSelector.obtainMultipleResult(data)
            if (selectList.size != 0 && selectList[0].isCut) {
                SharePrefUtil.setBoardImg(this, boardId, selectList[0].cutPath)
                loadBoardImg()
            }
        }
    }
}